lib/Doctrine/DBAL/Driver/AbstractMySQLDriver.php

```diff
@@ -123,35 +125,72 @@ public function convertException($message, DriverException $exception)
+     * @throws DBALException
      */
-    public function createDatabasePlatformForVersion($version)
+    public function createDatabasePlatformForVersion($version): AbstractPlatform
     {
-        if ( ! preg_match('/^(?P<major>\d+)(?:\.(?P<minor>\d+)(?:\.(?P<patch>\d+))?)?/', $version, $versionParts)) {
+        if (false !== stripos($version, 'mariadb')) {
+            $versionNumber = $this->getMariaDbMysqlVersionNumber($version);
+            if (version_compare($versionNumber, '10.2.7', '>=')) {
+                return new MariaDb1027Platform();
+            }
+        } else {
+            $versionNumber = $this->getOracleMysqlVersionNumber($version);
+            if (version_compare($versionNumber, '5.7.9', '>=')) {
+                return new MySQL57Platform();
+            }
+        }
+
+        return $this->getDatabasePlatform();
+    }
+
+    /**
+     * Get a normalized 'version number' from the server string
+     * returned by Oracle MySQL servers.
+     *
+     * @param string $versionString Version string returned by the driver, i.e. '5.7.10'
+     * @throws DBALException
+     */
+    private function getOracleMysqlVersionNumber(string $versionString): string
+    {
+        if (!preg_match('/^(?P<major>\d+)(?:\.(?P<minor>\d+)(?:\.(?P<patch>\d+))?)?/', $versionString, $versionParts)) {
             throw DBALException::invalidPlatformVersionSpecified(
-                $version,
+                $versionString,
                 '<major_version>.<minor_version>.<patch_version>'
             );
         }
-
-        if (false !== stripos($version, 'mariadb')) {
-            return $this->getDatabasePlatform();
-        }
-
         $majorVersion = $versionParts['major'];
-        $minorVersion = isset($versionParts['minor']) ? $versionParts['minor'] : 0;
-        $patchVersion = isset($versionParts['patch']) ? $versionParts['patch'] : null;
+        $minorVersion = $versionParts['minor'] ?? 0;
+        $patchVersion = $versionParts['patch'] ?? null;
 
         if ('5' === $majorVersion && '7' === $minorVersion && null === $patchVersion) {
             $patchVersion = '9';
         }
 
-        $version = $majorVersion . '.' . $minorVersion . '.' . $patchVersion;
+        return $majorVersion . '.' . $minorVersion . '.' . $patchVersion;
+    }
 
-        if (version_compare($version, '5.7.9', '>=')) {
-            return new MySQL57Platform();
+    /**
+     * Detect MariaDB server version, including hack for some mariadb distributions
+     * that starts with the prefix '5.5.5-'
+     *
+     * @param string $versionString Version string as returned by mariadb server, i.e. '5.5.5-Mariadb-10.0.8-xenial'
+     * @throws DBALException
+     */
+    private function getMariaDbMysqlVersionNumber(string $versionString): string
+    {
+        $version = str_replace('5.5.5-', '', $versionString);
+
+        if (!preg_match('/^(mariadb-)?(?P<major>\d+)\.(?P<minor>\d+)\.(?P<patch>\d+)/', strtolower($version), $versionParts)) {
```

Comment:
Looks like an almost complete copy-paste of the method above. Can/should be this logic generalized?

Reply:
I would love to... but not without adding more complexity. As covered by the existing tests[here](https://github.com/belgattitude/dbal/blob/248a05abaf72c5bc4488549a0bd143c124c66351/tests/Doctrine/Tests/DBAL/Functional/Schema/MySqlSchemaManagerTest.php#L377). Mysql can be 5.7, 5.7.1, 6?. Meaning only major version is required. Mariadb have two variants: one will full semver (10.0.15-MariaDB-1~wheezy) the other one prefixed by '5.5.5-'  (5.5.5-MariaDB-10.2.8+maria~xenial-log, 5.5.5-10.1.25-MariaDB). Note: the 5.5.5- prefix will probably live long in official maria release. Don't understand everything, but they made it on purpose to hack a limitation in the replication protocol (version is stored on x bytes).

Reply:
Hey @morozov , thanks for getting back on this. Yes it's more clear now. I don't specially like the way the mysql version detection was done. Both minor and patch versions are optional and covered by tests like:
```php
array('5.7', 'Doctrine\DBAL\Platforms\MySQL57Platform'),
array('6', 'Doctrine\DBAL\Platforms\MySQL57Platform'),
array('5.7.10', 'Doctrine\DBAL\Platforms\MySQL57Platform'),
```
Meaning the mariadb will have to do the same. For me it's much more safer to regexp with a full spec. What happens if suddently they package mariadb as '5.5.5-mariadb18-10.2.8-MariadDb-5.7~wheezy) ? Not sure the mysql regexp `/^(?P<major>\d+)(?:\.(?P<minor>\d+)(?:\.(?P<patch>\d+))?)?/`will match 10.2.8 instead of 18 or 5.7)... Of course it's hypothetical, but I've seen a lot of different patterns in mariadb.  Let me know, if you still convinced about having only one regexp, I can do. But not sure how it will be valid.

But I'll try to figure out, very tempting to have onoly one regexp of course.